### PR TITLE
MM-36244 - Generic upgrade banner/flow; Upgrade banner/flow for playbook dashboard

### PR DESCRIPTION
**Pull request by cpoile**
#### Summary
- For ease of review I split the commits into mostly independent chunks.
- Genericized the upgrade banner with upgrade flow (derived from the rhs timeline banner)
- Converted the rhs timeline banner to use this, and using it for the playbook dashboard
- Tested cloud admin, cloud user, self-hosted user: success and failure; could only test self-hosted admin failure
- Test self-hosted admin success only once, because after the first licence it wouldn't give me another. Will work on that.
- @justinegeffen I'll need copy for the notify admin message (I'll highlight that below), and a review of the copy in the screencaps below
- @abhijit-singh in order to make the upgrade banner generic, I chose to follow the way the rhs timeline presented feedback to the user on success and failure--not a modal, but in the banner itself. 
- @abhijit-singh @itao The trial request success modal (for this and all the other upgrade flows) might be out of scope for this epic (we're running into time constraints).

Where we are:
- [x] Backend stats, rough frontend. 
- [x] New navigation model (between list, landing, edit, messaging) 
- [x] UI tweaks to bring into 100% in line with designs
- [x] New: license checked and upgrade bannered  <-- We are here
- [ ] New: Add "Run playbook" button on playbook dashboard screen
- [ ] New: Improve graph searching (pill to show search filter, cursor) 
- [ ] New: Trial request success modal for all upgrade paths

Cloud admin pre:
image

Cloud admin post (success):
image

Cloud user / Self-hosted user pre:
image

Cloud user / Self-hosted user post (success):
image

Cloud user / Self-hosted user post (failure):
image

Self-hosted admin pre:
image

Self-hosted admin post (success): 
TODO: Need to test. But, it should immediately change into working stats view. 

Self-hosted admin post (failure):
image

#### Ticket Link
- https://mattermost.atlassian.net/browse/MM-36244

#### Checklist
- [x] Telemetry updated
- [x] Gated by experimental feature flag
- [ ] ~~Unit tests updated~~

### CODE REVIEW

File: server/bot/poster.go [hunk 1]
```diff
@@ -159,6 +159,10 @@ func (b *Bot) NotifyAdmins(messageType, authorUserID string, isTeamEdition bool)
 		message = fmt.Sprintf("@%s requested access to the timeline in Incident Collaboration.", author.Username)
 		title = "Keep a complete record of the playbook run timeline"
 		text = "The playbook run timeline automatically tracks key events and messages in chronological order so that they can be traced and reviewed afterwards. Teams use timeline to perform retrospectives and extract lessons for the next time that they run the playbook."
+	case "start_trial_to_access_playbook_dashboard":
+		message = fmt.Sprintf("@%s requested access to the playbook dashboard in Incident Collaboration.", author.Username)
+		title = "View playbook statistics in the Playbook Dashboard"
+		text = "The Playbook Dashboard shows trends for total runs, active runs, and participants involved in each type of playbook."
```
Comment (cpoile) [inline server/bot/poster.go:165]: @justinegeffen This is the copy used to notify the sys admin

File: webapp/src/components/upgrade_banner.tsx [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+type HandlerType = undefined | (() => (Promise<void> | void));
+
+interface Props {
+    illustration: JSX.Element;
+    successIllustration: JSX.Element;
+    errorIllustration: JSX.Element;
+    titleText: string;
+    helpText: string;
+    notificationType: AdminNotificationType;
+    upgradeWrapperSC: StyledComponent<'div', any>;
```
Comment (crspeller) [inline webapp/src/components/upgrade_banner.tsx:36]: Do we have to be specific with these types? Or just use JSX.Element?

File: webapp/src/components/upgrade_banner.tsx [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+enum ActionState {
+    Uninitialized,
+    Loading,
+    Error,
+    Success,
+}
+
+type HandlerType = undefined | (() => (Promise<void> | void));
+
+interface Props {
```
Comment (crspeller) [inline webapp/src/components/upgrade_banner.tsx:29]: Also is this level of customization on the individual areas necessary? Are they that different? Should they be?

**Review by crspeller — APPROVED** [review]
Non blocking questions. Thanks for doing this!

File: webapp/src/components/upgrade_banner.tsx (hunk 2 above)
Comment (cpoile) [reply webapp/src/components/upgrade_banner.tsx:36]: If we use JSX.Element, we can't treat it like a StyledComponent -- try it and you'll see the error. Unless there's an easy way to use it just like a SC? (Open to whatever works best)

File: webapp/src/components/upgrade_banner.tsx (hunk 3 above)
Comment (cpoile) [reply webapp/src/components/upgrade_banner.tsx:29]: Yep, as long as you want control over the styling of the elements (which I did for the backstage banner and text). I only genericized what I had to. I bet we could make the designs a bit more standard, but that would need more work and back and forth with Abhijit.

File: server/bot/poster.go [hunk 4]
```diff
@@ -159,6 +159,10 @@ func (b *Bot) NotifyAdmins(messageType, authorUserID string, isTeamEdition bool)
 		message = fmt.Sprintf("@%s requested access to the timeline in Incident Collaboration.", author.Username)
 		title = "Keep a complete record of the playbook run timeline"
 		text = "The playbook run timeline automatically tracks key events and messages in chronological order so that they can be traced and reviewed afterwards. Teams use timeline to perform retrospectives and extract lessons for the next time that they run the playbook."
+	case "start_trial_to_access_playbook_dashboard":
+		message = fmt.Sprintf("@%s requested access to the playbook dashboard in Incident Collaboration.", author.Username)
```
Comment (agarciamontoro) [inline server/bot/poster.go:163]: We are removing the "in Incident Collaboration" part. See https://github.com/mattermost/mattermost-plugin-incident-collaboration/pull/611/files
```suggestion
		message = fmt.Sprintf("@%s requested access to the playbook dashboard.", author.Username)
```

File: server/bot/poster.go (hunk 1 above)
Comment (agarciamontoro) [reply server/bot/poster.go:165]: @cpoile, Justine is on vacation, so we should either ask Carrie for a review or just ship this and review later.

**Review by agarciamontoro — APPROVED** [review]
This is fantastic! Thank you for the work on the common component. The only missing piece is the telemetry for this. Take a look to [this commit](https://github.com/mattermost/mattermost-plugin-incident-collaboration/pull/616/commits/bc7f9694a44060b076550d6f441f02d43f29fe0e) to see the needed changes.

Approving now to unblock the PR :tada:

**Review by abhijit-singh — APPROVED** [review]
Looks good @cpoile!

One small thing that I just noticed - can we make the background color of the page white (center-channel-bg)?

File: webapp/src/components/upgrade_banner.tsx (hunk 3 above)
Comment (crspeller) [reply webapp/src/components/upgrade_banner.tsx:90]: Talked about this and above offline. I pushed a commit fixing both of these concerns and making it easier to add this to more places in the future.

File: webapp/src/components/upgrade_banner.tsx [hunk 5]
```diff
@@ -188,11 +229,11 @@ const Button = (props: ButtonProps) => {
     if (props.actionState === ActionState.Error) {
         if (props.isCurrentUserAdmin) {
             return (
-                <ButtonSC
+                <PrimaryButton
```
Comment (cpoile) [inline webapp/src/components/upgrade_banner.tsx:232]: @crspeller @abhijit-singh Wants this to be a `Secondary` button for the playbook dashboard banner. I'm not sure what you need on yours. Can we assume vertical is a `Primary` button, non-vertical is `Secondary`?

Comment (crspeller) [reply webapp/src/components/upgrade_banner.tsx:232]: @abhijit-singh Are you sure? I don't see why we wouldn't be consistent.

Comment (cpoile) [reply webapp/src/components/upgrade_banner.tsx:232]: @crspeller I see two options: pass in a buttonSC, or a `secondaryButton` flag. Can you think of another? Any preference?

Comment (crspeller) [reply webapp/src/components/upgrade_banner.tsx:232]: @cpoile I would go with the flag. 2/5

Comment (cpoile) [reply webapp/src/components/upgrade_banner.tsx]: done!

**Pull request merged**